spock-core/src/main/java/org/spockframework/mock/runtime/IMockMaker.java

```diff
@@ -0,0 +1,131 @@
+ */
+@Beta
+@ThreadSafe
+public interface IMockMaker {
+  /**
+   * The system property, which allows to override the priority of the {@link IMockMaker}.
+   *
+   * <p>Please note that the {@link MockMakerId#dynamicProxy} has
+   * always priority over every other {@link IMockMaker}.
+   */
+  String MOCK_MAKER_PROPERTY = "org.spockframework.mock.runtime.mockmaker";
+
+  /**
+   * Returns the ID of this {@link IMockMaker}.
+   * The ID must be unique and should be in the form of {@code your-name-mock-maker}.
+   *
+   * @return the ID
+   */
+  String getId();
+
+  /**
+   * @return The priority in which the {@link IMockMaker}s are applied.
+   */
+  int getPriority();
+
+  /**
+   * Creates a new mock/stub/spy instance for the passed {@link IMockCreationSettings}
+   *
+   * @param settings the settings
+   * @return the new mock instance
+   * @throws CannotCreateMockException if this {@link IMockMaker} can't create such a mock.
+   */
+  Object makeMock(IMockCreationSettings settings) throws CannotCreateMockException;
+
+  ITypeMockability isTypeMockable(IMockCreationSettings settings);
+
+  /**
+   * Returns information about a mock object, or {@code null} it the object is no mock.
+   *
+   * <p>Implementation hint:
+   * An {@link IMockMaker} must provide the means to identify if a certain object is a mock, and need to return an attached {@link IMockObject}.
+   * If your mock class implement {@link org.spockframework.mock.ISpockMockObject}, you can simply return {@code null}.</p>
+   * </p>
+   *
+   * @param object a mock object
+   * @return information about the mock object or {@code null}
+   */
+  default IMockObject asMockOrNull(Object object) {
+    return null;
+  }
+
+  default void throwExceptionIfNotMockable(IMockCreationSettings settings) {
+    ITypeMockability typeMockable = isTypeMockable(settings);
+    if (!typeMockable.mockable()) {
+      throw new CannotCreateMockException(settings.getMockType(), ". " + typeMockable.nonMockableReason());
+    }
+  }
+
+  @Immutable
+  @FunctionalInterface
+  @Beta
+  interface ITypeMockability {
+    ITypeMockability MOCKABLE = new ITypeMockability() {
+      @Override
+      public boolean mockable() {
+        return true;
+      }
+
+      @Override
+      public String nonMockableReason() {
+        return "";
+      }
+    };
+
+    /**
+     * @return {@code true} if the type is mockable.
+     */
+    default boolean mockable() {
+      return false;
+    }
+
+    /**
+     * @return the reason why the type is not mockable, otherwise empty {@link String}.
+     */
+    String nonMockableReason();
+  }
+
+  @Beta
+  interface IMockCreationSettings {
+
+    @Nullable
+    String getMockMakerId();
+
+    Class<?> getMockType();
+
+    List<Class<?>> getAdditionalInterface();
+
+    @Nullable
+    List<Object> getConstructorArgs();
+
+    IProxyBasedMockInterceptor getMockInterceptor();
+
+    ClassLoader getClassLoader();
+
+    boolean isUseObjenesis();
```

Comment:
Maybe, we should forward any unknown map-parameters, or add some explicit `mockMakerArgs` option.
Not sure if we should add it now or YAGNI?

Reply:
I guess that would make it more future proof, and the cost to do that is not much.
But then I would also remove the ` isUseObjenesis()` and move it into the `mockMakerArgs´ because the property is strange right now anyway.

Reply:
I have added a type-safe way to define `IMockMakerSettings` and pass them to the `IMockMaker ` instance.
I think this is the more elegant way, because then the developer of an `IMockMaker` can provide own APIs instead of a plain argument map.